include/secp256k1.h

```diff
@@ -701,6 +708,188 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_combine(
+ *  quadratic residue), otherwise it is negative.
+ *
+ *  Returns: 1 if the public key was successfully converted
+ *           0 otherwise
+ *
+ *  Args:           ctx: pointer to a context object
+ *  Out:         pubkey: pointer to a public key object for placing the
+ *                       converted public key (cannot be NULL)
+ *  In: xonly_pubkey: pointer to an x-only public key that is converted
+ *                    (cannot be NULL)
+ *                 sign: sign bit of the resulting public key (can be NULL)
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_xonly_pubkey_to_pubkey(
+    const secp256k1_context* ctx,
+    secp256k1_pubkey *pubkey,
+    const secp256k1_xonly_pubkey *xonly_pubkey,
+    int sign
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Tweak the private key of an x-only pubkey by adding a tweak to it. The public
+ *  key of the resulting private key will be the same as the output of
+ *  secp256k1_xonly_pubkey_tweak_add called with the same tweak and corresponding
+ *  input public key.
+ *
+ *  If the public key corresponds to a positive point, tweak32 is added to the
+ *  seckey (modulo the group order). If the public key corresponds to a
+ *  negative point, tweak32 is added to the negation of the seckey (modulo the
+ *  group order).
+ *
+ *  Returns: 1 if the tweak was successfully added to seckey
+ *           0 if the tweak was out of range or the resulting private key would be
+ *             invalid (only when the tweak is the complement of the private key) or
+ *             seckey is 0.
+ *
+ *  Args:      ctx: pointer to a context object, initialized for signing (cannot be NULL)
+ *  In/Out: seckey: pointer to a 32-byte private key
+ *  In:    tweak32: pointer to a 32-byte tweak
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_xonly_privkey_tweak_add(
+    const secp256k1_context* ctx,
+    unsigned char *seckey,
+    const unsigned char *tweak32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Tweak an x-only public key by adding tweak times the generator to it. Note that
+ *  the output is a secp256k1_pubkey and not a secp256k1_xonly_pubkey.
+ *  Returns: 1 if tweak times the generator was successfully added to pubkey
+ *           0 if the tweak was out of range or the resulting public key would be
+ *             invalid (only when the tweak is the complement of the corresponding
+ *             private key).
+ *
+ *  Args:           ctx: pointer to a context object initialized for validation
+ *                       (cannot be NULL)
+ *  Out:  output_pubkey: pointer to a public key object (cannot be NULL)
+ *  In: internal_pubkey: pointer to an x-only public key object to apply the
+ *                       tweak to (cannot be NULL)
+ *              tweak32: pointer to a 32-byte tweak (cannot be NULL)
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_xonly_pubkey_tweak_add(
```

Comment:
I'm in favor of high level operations which is why my other open pull requests have similar commitment APIs than what real-or-random suggests. For this PR I had a different mindset, because when I brought up high level functions you mentioned "ec_pubkey_tweak is perfect" which was a misunderstanding apparently.

Having tweak_add/tweak_add_verify functions working on xonly_pubkeys/bools is what I started out with, but then decided against that because 1) the output of tweak_add *is* a secp_pubkey, so returning a pubkey is a better separation of layers, 2) the user otherwise has to deal with two pieces of data of data 3) musig_pubkey_combine takes a tweak as well, and would either have to get an additional out argument with the sign or would work differently than tweak_add and return a pubkey.

However, if we want to add a high level API anyway, I'm happy with changing this again to the more hacky version if it's more straightforward to use in your taproot implementation.

Reply:
I see the burden of forcing the user to deal with two pieces of data in some cases, but I think it's justified. The BIPs do not mention any conversion from 32byte-pubkeys to point-pubkeys, only the other way around.

Reply:
Yeah, I think I changed my mind after seeing the complexity necessary to have both close-to-optimal performance and an intermediate-layer level API.

Reply:
Just to clarify, what are the opinions you changed your mind from and to?